File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/AnalogValueBase.vb
'*******************************************************************************************************
'  AnalogValueBase.vb - Analog value base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  02/18/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization

''' <summary>This class represents the common implementation of the protocol independent representation of an analog value.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class AnalogValueBase

    Inherits ChannelValueBase(Of IAnalogDefinition)
    Implements IAnalogValue

    Private m_value As Single

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize analog value
        m_value = info.GetSingle("value")

    End Sub

    Protected Sub New(ByVal parent As IDataCell)

        MyBase.New(parent)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal value As Single)
    Protected Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal value As Single)

        MyBase.New(parent, analogDefinition)

        m_value = value

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal unscaledValue As Int16)
    Protected Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal integerValue As Int16)

        MyClass.New(parent, analogDefinition, Convert.ToSingle(integerValue))

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    Protected Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, analogDefinition)
        ParseBinaryImage(Nothing, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected expose a Public Sub New(ByVal analogValue As IAnalogValue)
    Protected Sub New(ByVal analogValue As IAnalogValue)

        MyClass.New(analogValue.Parent, analogValue.Definition, analogValue.Value)

    End Sub

    Public Overridable Property Value() As Single Implements IAnalogValue.Value
        Get
            Return m_value
        End Get
        Set(ByVal value As Single)
            m_value = value
        End Set
    End Property

    Public Overridable Property IntegerValue() As Int16 Implements IAnalogValue.IntegerValue
        Get
            Return Convert.ToInt16(m_value)
        End Get
        Set(ByVal value As Int16)
            m_value = Convert.ToSingle(value)
        End Set
    End Property

    Default Public Overrides Property CompositeValue(ByVal index As Integer) As Single
        Get
            Return m_value
        End Get
        Set(ByVal value As Single)
            m_value = value
        End Set
    End Property

    Public Overrides ReadOnly Property CompositeValueCount() As Integer
        Get
            Return 1
        End Get
    End Property

    Public Overrides ReadOnly Property IsEmpty() As Boolean
        Get
            Return (m_value = 0)
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            If DataFormat = Phasors.DataFormat.FixedInteger Then
                Return 2
            Else
                Return 4
            End If
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = CreateArray(Of Byte)(BodyLength)

            If DataFormat = Phasors.DataFormat.FixedInteger Then
                EndianOrder.BigEndian.CopyBytes(IntegerValue, buffer, 0)
            Else
                EndianOrder.BigEndian.CopyBytes(m_value, buffer, 0)
            End If

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Integer)

        If DataFormat = Phasors.DataFormat.FixedInteger Then
            IntegerValue = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
        Else
            m_value = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex)
        End If

    End Sub

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize analog value
        info.AddValue("value", m_value)

    End Sub

    Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
        Get
            With MyBase.Attributes
                .Add("Analog Value (Floating Point)", Value)
                .Add("Analog Value (Integer)", IntegerValue)
            End With

            Return MyBase.Attributes
        End Get
    End Property

End Class

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/AnalogValueBase.cs
using System.Diagnostics;
using System;
//using TVA.Common;
using System.Collections;
using TVA.Interop;
using Microsoft.VisualBasic;
using TVA;
using System.Collections.Generic;
//using TVA.Interop.Bit;
using System.Linq;
using System.Runtime.Serialization;

//*******************************************************************************************************
//  AnalogValueBase.vb - Analog value base class
//  Copyright © 2008 - TVA, all rights reserved - Gbtc
//
//  Build Environment: VB.NET, Visual Studio 2008
//  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
//      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
//       Phone: 423/751-2827
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/18/2005 - J. Ritchie Carroll
//       Initial version of source generated
//
//*******************************************************************************************************

namespace PhasorProtocols
{
    /// <summary>This class represents the common implementation of the protocol independent representation of an analog value.</summary>
    [CLSCompliant(false), Serializable()]
    public abstract class AnalogValueBase : ChannelValueBase<IAnalogDefinition>, IAnalogValue
    {



        private float m_value;
        private bool m_valueAssigned;

        protected AnalogValueBase()
        {
        }

        protected AnalogValueBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {


            // Deserialize analog value
            m_value = info.GetSingle("value");
            m_valueAssigned = true;

        }

        protected AnalogValueBase(IDataCell parent)
            : base(parent)
        {


        }

        // Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal value As Single)
        protected AnalogValueBase(IDataCell parent, IAnalogDefinition analogDefinition, float value)
            : base(parent, analogDefinition)
        {


            m_value = value;
            m_valueAssigned = !float.IsNaN(value);

        }

        // Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal unscaledValue As short)
        protected AnalogValueBase(IDataCell parent, IAnalogDefinition analogDefinition, short integerValue)
            : this(parent, analogDefinition, (float)integerValue)
        {


        }

        // Derived classes are expected expose a Public Sub New(ByVal parent As IDataCell, ByVal analogDefinition As IAnalogDefinition, ByVal binaryImage As Byte(), ByVal startIndex As int)
        protected AnalogValueBase(IDataCell parent, IAnalogDefinition analogDefinition, byte[] binaryImage, int startIndex)
            : base(parent, analogDefinition)
        {

            ParseBinaryImage(null, binaryImage, startIndex);

        }

        // Derived classes are expected expose a Public Sub New(ByVal analogValue As IAnalogValue)
        protected AnalogValueBase(IDataCell parent, IAnalogDefinition analogDefinition, IAnalogValue analogValue)
            : this(parent, analogDefinition, analogValue.Value)
        {


        }

        public virtual float Value
        {
            get
            {
                return m_value;
            }
            set
            {
                m_value = value;
                m_valueAssigned = true;
            }
        }

        public virtual short IntegerValue
        {
            get
            {
                try
                {
                    return (short)m_value;
                }
                catch (OverflowException)
                {
                    return short.MinValue;
                }
            }
            set
            {
                m_value = (float)value;
                m_valueAssigned = true;
            }
        }

        public override float this[int index]
        {
            get
            {
                return m_value;
            }
            set
            {
                m_value = value;
                m_valueAssigned = true;
            }
        }

        public override int CompositeValueCount
        {
            get
            {
                return 1;
            }
        }

        public override bool IsEmpty
        {
            get
            {
                return !m_valueAssigned;
            }
        }

        protected override ushort BodyLength
        {
            get
            {
                if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
                {
                    return 2;
                }
                else
                {
                    return 4;
                }
            }
        }

        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];

                if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
                {
                    EndianOrder.BigEndian.CopyBytes(IntegerValue, buffer, 0);
                }
                else
                {
                    EndianOrder.BigEndian.CopyBytes(m_value, buffer, 0);
                }

                return buffer;
            }
        }

        protected override void ParseBodyImage(IChannelParsingState state, byte[] binaryImage, int startIndex)
        {

            if (DataFormat == PhasorProtocols.DataFormat.FixedInteger)
            {
                IntegerValue = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex);
            }
            else
            {
                m_value = EndianOrder.BigEndian.ToSingle(binaryImage, startIndex);
                m_valueAssigned = true;
            }

        }

        public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
        {

            base.GetObjectData(info, context);

            // Serialize analog value
            info.AddValue("value", m_value);

        }

        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Analog Value (Floating Point)", Value.ToString());
                baseAttributes.Add("Analog Value (Integer)", IntegerValue.ToString());

                return baseAttributes;
            }
        }

    }
}
